rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+  * Search (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+* Analytics Service
+  * Query (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+
+The Indexing Service and Eventing Services are not included as they are not
+network-available services for applications to make requests from.
+
+As with the Service Experience Counters above, it may be incorrect to try to
+compare between deployments.  For example, the query complexity may differ
+significantly or the data size and use of features like compression may vary
+significantly between two applications.
+
+# API and Implementation Details
+
+It should be possible to reuse existing Meter implementation, that according to
+RFC-67 (Extended SDK Observability) already builds histograms for request
+latencies.
+
+In addition to latencies, which would be used for reporting histograms, the
+default Meter should also maintain the following counters:
+
+| Metric | Description |
+| :---- | :---- |
+| sdk\_*{service}*\_r\_utimedout | Unambiguous timeout |
+| sdk\_*{service}*\_r\_atimedout | Ambiguous timeout |
+| sdk\_*{service}*\_r\_canceled | Canceled |
+| sdk\_*{service}*\_r\_total | Total number of operations |
+
+Where the service is a part of the visible counter name, and hostname and
+bucket must be stored and exported as a meta data.
+
+## App Telemetry Activation
+
+The collection and reporting of application telemetry should be activated by
+either of the following mechanisms:
+
+* presence of the `appTelemetry` property of the node in `nodexExt` entry of
+  the configuration (NOTE: the exact name and semantics of the value might be
+  defined by the server team later). In this case the SDK would either use
+  hardcoded path or extract one from the JSON configuration and use for and
+  build the endpoint address the websocket connection.
+
+* presence of the `app_telemetry_endpoint` property in the connection option,
+  or similar optiion in the `ClusterOptions`:
+
+      couchbases://abc.cloud.couchbase.com?app_telemetry_endpoint=ws://localhost:9090/app_telemetry
+
+      ClusterOptions options;
+      options.appTelemetryEndpoint("ws://localhost:9090/app_telemetry");
+
+  This endpoint must not have authentication, and should support WebSocket
+  protocol.
+
+## Network Interacation
+
+Once the SDK determined the endpoint it should establish HTTP connection and
+upgrade it to WebSocket.
+
+The server allows to close connection at any point of time.
+
+Once the server closes connection, the SDK should pick next available endpoint.
+Both initial and next endpoints must be randomized in case of more than one
+endpoint configured.
+
+The SDK must wait for backoff interval before connection attempt to avoid
+stressing the server. The SDK might use `app_telemetry_backoff` option to allow
+to change the backoff value.
+
+The SDK must either implement or enable WebSocket PING frame, and respond to
```

Comment:
To detect dead connections from SDK to server, should the SDK ping the server if the server hasn't sent a request in some time period, and drop/reestablish the connection if the server does not respond within some time period?

Reply:
This seems like a good idea, but maybe the ping is a bit of additional complexity... Maybe we just establish a sensible maximum interval with the server team that we'd expect to see a poll within (a few minutes?), and drop+reestablish based on that?

Reply:
I have updated RFC, and added line that the SDK might initiate PING, and reconnect after `app_telemetry_backoff`. But I don't think it should be mandatory for the implementation.